(In review)Hanson/add examples to plotly (#736)
reflex-dev/reflex-web: docs/library/graphing/plotly.md

```md alert info
# When integrating Plotly graphs into your UI code, note that the method for displaying the graph differs from a regular Python script. Instead of using `fig.show()`, use `rx.plotly(data=fig)` within your UI code to ensure the graph is properly rendered and displayed within the user interface
```
## Basic Example
Let's create a line graph of life expectancy in Canada.

```python demo exec
import plotly.express as px

df = px.data.gapminder().query("country=='Canada'")
fig = px.line(df, x="year", y="lifeExp", title='Life expectancy in Canada')  

def line_chart():
    return rx.center(
        rx.plotly(data=fig),
    )
```
## 3D graphing example
Let's create a 3D surface plot of Mount Bruno. 

```python demo exec
import plotly.graph_objects as go
import pandas as pd

# Read data from a csv
z_data = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/api_docs/mt_bruno_elevation.csv')

fig = go.Figure(data=[go.Surface(z=z_data.values)])
fig.update_traces(contours_z=dict(show=True, usecolormap=True,
                                  highlightcolor="limegreen", project_z=True))
fig.update_layout(
    scene_camera_eye=dict(x=1.87, y=0.88, z=-0.64),

>     width=500, height=500,

Reviewer: should we be setting width and height here?

        rx.plotly(data=fig),
    )
```

## Plot as State Var
Set plot as a State var when the plot need to be updated during run time. 

```python demo exec
import plotly.express as px
class PlotlyState(rx.State):
    df = px.data.gapminder().query(f"country=='Canada'")
    figure = px.line(
        df,
        x="year",
        y="lifeExp",
        title="Life expectancy in Canada",
    )

    def set_selected_country(self, country):
        self.df = px.data.gapminder().query(f"country=='{country}'")
        self.figure = px.line(
            self.df,
            x="year",
            y="lifeExp",
            title=f"Life expectancy in {country}",
        )

def line_chart_with_state():
    return rx.vstack(
        rx.select(
            ['China', 'France', 'United Kingdom', 'United States', 'Canada'],
            default_value="United States",
            on_change=PlotlyState.set_selected_country,
        ),
        rx.plotly(data=PlotlyState.figure),
    )
```

## Adding Styles and Layouts
Use `update_layout()` method to update the layout of your chart. Checkout [Plotly Layouts](https://plotly.com/python/reference/layout/) for all layouts props. 

```md alert info

> Note that the width and height props is currently unsupported. 

Reviewer: ```suggestion
Note that the width and height props are not recommended to ensure the plot remains size responsive to its container. 
```

> Note that the width and height props are not recommended to ensure the plot remains size responsive to its container. 

Reviewer 2: It is not clear what these width and height props are relating to in this example, is it the update_layout or is it rx.plotly

df = px.data.gapminder().query("country=='Canada'")
fig_1 = px.line(
    df,
    x="year",
    y="lifeExp",
    title="Life expectancy in Canada",
)
fig_1.update_layout(
    title_font_size = 25,
    title_x = 0.5, # at the center",
    plot_bgcolor = "#c3d7f7",
    paper_bgcolor = "#e0e0e0",
)
def add_styles():
    return rx.center(
        rx.plotly(data=fig_1),
    )
```

## Dimension and Resizing
The auto-resizing is enabled by default (set to True). The size of the plot will be responsive according to the screen size. To disable auto-resizing, you need to set `use_resize_handler` to False. You can size your Plotly chart using the width and height props, just like any other Reflex component.

```md alert info
It is recommended to set the width and height in the `rx.plotly` component instead of using the `update_layout()` method. The latter will disable auto-resizing.
```

```python demo exec
df = px.data.gapminder().query("country=='Canada'")
fig_1 = px.line(
    df,
    x="year",
    y="lifeExp",
    title="Life expectancy in Canada",
)
fig_1.update_layout(
    title_font_size = 25,
    title_x = 0.5,
    plot_bgcolor = "#c3d7f7",
    paper_bgcolor = "#e0e0e0",
)

> def resizing():

Reviewer 2: This resize example does not actually seem to resize when I change the size of the screen (at least it does not in safari)

import pandas as pd
import plotly.express as px
import plotly.graph_objects as go
from pcweb import constants
```

> Plotly is a graphing library that can be used to create interactive graphs.

Reviewer 2: Somewhere in this page we need a clear explanation of how to resize the graph in your app. If it is using `fig.update_layout` or if it is just using the `width` and `height` props.

## Basic Example
Let's create a line graph of life expectancy in Canada.

```python demo exec
import plotly.express as px

df = px.data.gapminder().query("country=='Canada'")
fig = px.line(df, x="year", y="lifeExp", title='Life expectancy in Canada')  

def line_chart():
    return rx.center(
        rx.plotly(data=fig),
    )
```

> ```md alert info

Reviewer 2: I would add this right under line 16 (Plotly is a graphing library that can be used to create interactive graphs.) so that is is super obvious to new users and cannot be missed.

fig = go.Figure(data=[go.Surface(z=z_data.values)])
fig.update_traces(contours_z=dict(show=True, usecolormap=True,
                                  highlightcolor="limegreen", project_z=True))
fig.update_layout(
    scene_camera_eye=dict(x=1.87, y=0.88, z=-0.64),
    margin=dict(l=65, r=50, b=65, t=90)
)

def mountain_surface():
    return rx.center(
        rx.plotly(data=fig),
    )
```

## Plot as State Var

> The figure need to be set as a state var if it needs to be updated during run time. 

Reviewer: ```suggestion
If the figure is set as a state var, it can be updated during run time. 
```

```python demo exec
import plotly.express as px
class PlotlyState(rx.State):
    df = px.data.gapminder().query(f"country=='Canada'")
    figure = px.line(
        df,
        x="year",
        y="lifeExp",
        title="Life expectancy in Canada",
    )

    def set_selected_country(self, country):
        self.df = px.data.gapminder().query(f"country=='{country}'")
        self.figure = px.line(
            self.df,
            x="year",
            y="lifeExp",
            title=f"Life expectancy in {country}",
        )

def line_chart_with_state():
    return rx.vstack(
        rx.select(
            ['China', 'France', 'United Kingdom', 'United States', 'Canada'],

>             default_value="United States",

Reviewer: ```suggestion
            default_value="Canada",
```

The figure is initialized with Canada data, so the `default_value` for the select should match the initial state of the graph (even though it won't stick if the page is refreshed)

            on_change=PlotlyState.set_selected_country,
        ),
        rx.plotly(data=PlotlyState.figure),
    )
```

## Adding Styles and Layouts
Use `update_layout()` method to update the layout of your chart. Checkout [Plotly Layouts](https://plotly.com/python/reference/layout/) for all layouts props. 

```md alert info
Note that the width and height props are not recommended to ensure the plot remains size responsive to its container. The size of plot will be determined by it's outer container. 
```

```python demo exec
df = px.data.gapminder().query("country=='Canada'")
fig_1 = px.line(
    df,
    x="year",
    y="lifeExp",
    title="Life expectancy in Canada",
)
fig_1.update_layout(
    title_x=0.5,
    plot_bgcolor="#c3d7f7",
    paper_bgcolor="rgba(128, 128, 128, 0.1)",
    showlegend=True,

>     title_font_family="Open Sans",

Reviewer: On my machine, the plot does not actually render with "Open Sans" title font. This looks like some default serif font.

![image](https://github.com/reflex-dev/reflex-web/assets/1524005/ab34e634-e666-471a-84db-c64dd6b6178e)

Author: Hmmm, does it have to do with need to have  Open Sans font installed on your system? If not it will render as default.

fig = go.Figure(data=[go.Surface(z=z_data.values)])
fig.update_traces(contours_z=dict(show=True, usecolormap=True,
                                  highlightcolor="limegreen", project_z=True))
fig.update_layout(
    scene_camera_eye=dict(x=1.87, y=0.88, z=-0.64),
    width=500, height=500,
    margin=dict(l=65, r=50, b=65, t=90)
)

def mountain_surface():
    return rx.center(
        rx.plotly(data=fig),
    )
```

> ## Plot as State Var

Reviewer 2: For this graph we should add more graph types and enable them to be shifted through using a select. A bar chart and a scatter plot as well.

Author: WIP For next release.
